openshift/dynamic/client.py

```diff
@@ -403,13 +324,223 @@ def __getattr__(self, name):
 
 
-class ResourceContainer(object):
+class Discoverer(object):
+    """
+        A convenient container for storing discovered API resources. Allows
+        easy searching and retrieval of specific resources.
+
+        Subclasses implement the abstract methods with different loading strategies.
+    """
+
+    def __init__(self, client):
+        self.client = client
+        self._load_server_info()
+
+    @abstractproperty
+    def api_groups(self):
+        pass
+
+    @abstractmethod
+    def get(self, prefix=None, group=None, api_version=None, kind=None, **kwargs):
+        pass
+
+    @abstractmethod
+    def search(self, prefix=None, group=None, api_version=None, kind=None, **kwargs):
+        pass
+
+    @property
+    def version(self):
+        return self.__version
+
+    def _load_server_info(self):
+        self.__version = {'kubernetes': load_json(self.client.request('get', '/version'))}
+        try:
+            self.__version['openshift'] = load_json(self.client.request('get', '/version/openshift'))
+        except ApiException:
+            pass
+
+    def get_resources_for_api_version(self, prefix, group, version, preferred):
+        """ returns a dictionary of resources associated with provided (prefix, group, version)"""
+
+        resources = {}
+        subresources = {}
+
+        path = '/'.join(filter(None, [prefix, group, version]))
+        resources_response = load_json(self.client.request('GET', path))['resources']
+
+        resources_raw = list(filter(lambda resource: '/' not in resource['name'], resources_response))
+        subresources_raw = list(filter(lambda resource: '/' in resource['name'], resources_response))
+        for subresource in subresources_raw:
+            resource, name = subresource['name'].split('/')
+            if not subresources.get(resource):
+                subresources[resource] = {}
+            subresources[resource][name] = subresource
+
+        for resource in resources_raw:
+            # Prevent duplicate keys
+            for key in ('prefix', 'group', 'api_version', 'client', 'preferred'):
+                resource.pop(key, None)
+
+            resources[resource['kind']] = Resource(
+                prefix=prefix,
+                group=group,
+                api_version=version,
+                client=self.client,
+                preferred=preferred,
+                subresources=subresources.get(resource['name']),
+                **resource
+            )
+        return resources
+
+
+class LazyDiscoverer(Discoverer):
     """ A convenient container for storing discovered API resources. Allows
-        easy searching and retrieval of specific resources
+        easy searching and retrieval of specific resources.
+
+        Resources for the cluster are loaded lazily.
     """
 
-    def __init__(self, resources):
-        self.__resources = resources
+    class ResourceGroup(object):
+        def __init__(self, preferred, resources={}):
```

Comment:
You probably don't want to use a mutable default value for resources, I would default it to None and then do `resources = resources or {}`